ledger/ledger-on-posix-filesystem/src/main/scala/com/daml/ledger/on/filesystem/posix/FileSystemLedgerReaderWriter.scala

```diff
@@ -0,0 +1,216 @@
+
+  private val lock = new FileSystemLock(lockPath)
+
+  private val engine = Engine()
+
+  private val dispatcher: Dispatcher[Index] =
+    Dispatcher(
+      "posix-filesystem-participant-state",
+      zeroIndex = StartOffset,
+      headAtInitialization = StartOffset,
+    )
+
+  private val randomNumberGenerator = new Random()
+
+  override def checkHealth(): HealthStatus = Healthy
+
+  override def close(): Unit = {
+    dispatcher.close()
+  }
+
+  override def retrieveLedgerId(): LedgerId = ledgerId
+
+  override def events(offset: Option[Offset]): Source[LedgerRecord, NotUsed] =
+    dispatcher
+      .startingAt(
+        offset
+          .map(_.components.head.toInt)
+          .getOrElse(StartOffset),
+        OneAfterAnother[Index, immutable.Seq[LedgerRecord]](
+          (index: Index, _) => index + 1,
+          (index: Index) => retrieveLogEntry(index).map(immutable.Seq(_))
+        )
+      )
+      .mapConcat {
+        case (_, updates) => updates
+      }
+
+  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] = {
+    val submission = Envelope
+      .openSubmission(envelope)
+      .getOrElse(throw new IllegalArgumentException("Not a valid submission in envelope"))
+    lock.run {
+      for {
+        stateInputStream <- Future.sequence(
+          submission.getInputDamlStateList.asScala.toVector
+            .map(key => readState(key).map(key -> _)))
+        stateInputs: Map[DamlStateKey, Option[DamlStateValue]] = stateInputStream.toMap
+        entryId = allocateEntryId()
+        (logEntry, stateUpdates) = KeyValueCommitting.processSubmission(
+          engine,
+          entryId,
+          currentRecordTime(),
+          LedgerReader.DefaultTimeModel,
+          submission,
+          participantId,
+          stateInputs
+        )
+        _ = verifyStateUpdatesAgainstPreDeclaredOutputs(stateUpdates, entryId, submission)
+        newHead <- appendLog(entryId, Envelope.enclose(logEntry))
+        _ <- updateState(stateUpdates)
+      } yield {
+        dispatcher.signalNewHead(newHead)
+        SubmissionResult.Acknowledged
+      }
+    }
+  }
+
+  private def verifyStateUpdatesAgainstPreDeclaredOutputs(
+      actualStateUpdates: Map[DamlStateKey, DamlStateValue],
+      entryId: DamlLogEntryId,
+      submission: DamlSubmission
+  ): Unit = {
+    val expectedStateUpdates = KeyValueCommitting.submissionOutputs(entryId, submission)
+    if (!(actualStateUpdates.keySet subsetOf expectedStateUpdates)) {
+      val unaccountedKeys = actualStateUpdates.keySet diff expectedStateUpdates
+      sys.error(
+        s"CommitActor: State updates not a subset of expected updates! Keys [$unaccountedKeys] are unaccounted for!")
+    }
+  }
+
+  private def currentRecordTime(): Timestamp =
+    Timestamp.assertFromInstant(Clock.systemUTC().instant())
+
+  private def allocateEntryId(): DamlLogEntryId = {
+    val nonce: Array[Byte] = Array.ofDim(8)
```

Comment:
You may want to simply use a sequential log entry ID instead of a random one.

Reply:
Done; thanks for the changes that made this easy.